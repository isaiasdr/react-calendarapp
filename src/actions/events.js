import Swal from "sweetalert2";

import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";


export const eventStartAddNew = ( event ) => {
    return async ( dispatch, getState ) => {
        const { uid, name } = getState().auth;

        try {
            const response = await fetchWithToken('events', event, 'POST');
            const body = await response.json();

            if(body.ok) {
                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name
                };

                dispatch( eventAddNew(event) );
                Swal.fire({
                    title: 'Exito',
                    text: 'Evento guardado con exito',
                    icon: 'success'
                });
            }

            else {
                Swal.fire({
                    title: 'Error',
                    text: body.message,
                    icon: 'error'
                });
            }
        } catch (error) {
            console.log(error);
            Swal.fire({
                title: 'Error',
                text: 'error, pongase en contacto con el administrador',
                icon: 'error'
            });
        }
    }
}

const eventAddNew = (event) => ({
    type: types.eventAdded,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActiveEvent = () => ({
    type: types.eventClearActiveEvent
});

export const eventStartUpdate = (event) => {
    return async (dispatch) => {
        try {
            const response = await fetchWithToken(`events/${event.id}`, event, 'PUT');
            const body = await response.json();

            if(body.ok) {
                dispatch( eventUpdated(event) );
                Swal.fire({
                    title: 'Exito',
                    text: 'Evento actualizado con exito',
                    icon: 'success'
                });
            } else {
                Swal.fire({
                    title: 'Error',
                    text: body.message,
                    icon: 'error'
                });
            }

        } catch (error) {
            console.log(error);
            Swal.fire({
                title: 'Error',
                text: 'error, pongase en contacto con el administrador',
                icon: 'error'
            });
        }
    }
};

const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
});

export const eventStartDelete = () => {
    return async (dispatch, getState) => {
        const { activeEvent } = getState().calendar;
        const { id } = activeEvent;

        try {
            const response = await fetchWithToken(`events/${id}`, {}, 'DELETE');
            const body = await response.json();

            if(body.ok) {
                dispatch( eventDeleted() );
                Swal.fire({
                    title: 'Exito',
                    text: 'Evento eliminado con exito',
                    icon: 'success'
                });
            } else {
                Swal.fire({
                    title: 'Error',
                    text: body.message,
                    icon: 'error'
                });
            }

        } catch (error) {
            console.log(error);
            Swal.fire({
                title: 'Error',
                text: 'error, pongase en contacto con el administrador',
                icon: 'error'
            });
        }
    }
}

const eventDeleted = () => ({
    type: types.eventDeleted
});

export const eventSetNewEvent = (event) => ({
    type: types.eventSetNewEvent,
    payload: event
});

export const eventClearNewEvent = () => ({
    type: types.eventClearNewEvent
});

export const eventStartLoading = () => {
    return async ( dispatch ) => {
        try {
            const response = await fetchWithToken( 'events' );
            const body = await response.json();

            if(body.ok) {
                const events = body.events;

                dispatch( eventLoaded( prepareEvents(events) ) );
            }
                

            else {
                Swal.fire({
                    title: 'Error',
                    text: body.message,
                    icon: 'error'
                });
            }
        } catch (error) {
            console.log(error);
        }
    }
};

const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
});

export const eventClearAllLogout = () => ({
    type: types.eventClearAllLogout
});