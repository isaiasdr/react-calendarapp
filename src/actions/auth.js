import Swal from "sweetalert2";

import { types } from "../types/types"
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { eventClearAllLogout } from "./events";


export const startLogin = (email, password) => {
    return async ( dispatch ) => {
        const response = await fetchWithoutToken('auth', { email, password }, 'POST');
        const body = await response.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({
                uid: body.uid,
                name: body.name
            }));
        } 
        
        else {
            Swal.fire({
                title: 'Error',
                text: body.message,
                icon: 'error'
            });
        }
    };
};

export const startRegister = (name, email, password) => {
    return async ( dispatch ) => {

        try {
            const response = await fetchWithoutToken('auth/new', { email, name, password }, 'POST');
            const body = await response.json();

            if( body.ok ) {
                localStorage.setItem('token', body.token);
                localStorage.setItem('token-init-date', new Date().getTime() );

                dispatch( login({
                    uid: body.uid,
                    name: body.name
                }));
            } 
            
            else {
                Swal.fire({
                    title: 'Error',
                    text: body.message,
                    icon: 'error'
                });
            }
        } catch (error) {
            console.log(error);
            Swal.fire({
                title: 'Error',
                text: 'error, pongase en contacto con el administrador',
                icon: 'error'
            });
        }
    }
};

const login = ( user ) => ({
    type: types.authLogin,
    payload: user
});

export const startChecking = () => {
    return async ( dispatch ) => {
        const response = await fetchWithToken( 'auth/renew' );
        const body = await response.json();

        if( body.ok ) {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime() );

            dispatch( login({
                uid: body.uid,
                name: body.name
            }));
        }
        
        else {
            dispatch( checkFinish() );
        }
    };
}

const checkFinish = () => ({
    type: types.authCheckingFinish
});

export const startLogout = () => {
    return ( dispatch ) => {
        localStorage.clear();
        dispatch( eventClearAllLogout() );
        dispatch( logout() );
    }
}

const logout = () => ({
    type: types.authLogout
});