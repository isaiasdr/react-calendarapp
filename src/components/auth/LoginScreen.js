import React from 'react';
import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';
import validator from 'validator';

import { startLogin } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

import './login.css';


export const LoginScreen = () => {

    const initialForm = {
        email: 'testing1@test.com',
        password: '123456'
    };

    const [ formValues, handleInputChange ] = useForm( initialForm );

    const { email, password } = formValues;

    const dispatch = useDispatch();

    const handleLogin = (e) => {
        e.preventDefault();
        if (isValidForm())
            dispatch( startLogin(email, password) );
    };


    const isValidForm = () => {
        if (!validator.isEmail(email)) {
            Swal.fire({
                title: 'Error',
                text: 'El correo no tiene un formato valido',
                icon: 'error'
            })
            return false;
        }
            

        if (password.length < 6) {
            Swal.fire({
                title: 'Error',
                text: 'La contraseña debe tener al menos 6 caracteres',
                icon: 'error'
            });
            return false;
        }

        return true;
    }

    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-12 login-form-1">
                    <h3>Ingreso</h3>
                    <form onSubmit={ handleLogin }>
                        <div className="form-group">
                            <input 
                                type="text"
                                className="form-control"
                                placeholder="Correo"
                                name="email"
                                value={ email }
                                onChange={ handleInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña"
                                name='password'
                                value={ password }
                                onChange={ handleInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input 
                                type="submit"
                                className="btnSubmit"
                                value="Login" 
                            />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}
