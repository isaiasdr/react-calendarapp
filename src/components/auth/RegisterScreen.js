import React from 'react';
import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';
import validator from 'validator';

import { startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import './login.css';


export const RegisterScreen = () => {

    const initialForm = {
        email: '',
        name: '',
        password: '',
        password2: ''
    };

    const [ formValues, handleInputChange ] = useForm( initialForm );

    const { email, name, password, password2 } = formValues;

    const dispatch = useDispatch();

    const handleRegister = (e) => {
        e.preventDefault();
        if (isValidForm())
            dispatch( startRegister(name, email, password) );
    };

    const isValidForm = () => {
        if( name.length < 3 ) {
            Swal.fire({
                title: 'Error',
                text: 'El nombre debe tener al menos 3 caracteres',
                icon: 'error'
            });
            return false;
        }

        if (!validator.isEmail(email)) {
            Swal.fire({
                title: 'Error',
                text: 'El correo no tiene un formato valido',
                icon: 'error'
            });
            return false;
        }

        if (password.length < 6) {
            Swal.fire({
                title: 'Error',
                text: 'La contrase単a debe tener al menos 6 caracteres',
                icon: 'error'
            });
            return false;
        }

        if (password !== password2) {
            Swal.fire({
                title: 'Error',
                text: 'Las contrase単as no coinciden',
                icon: 'error'
            });
            return false;
        }

        return true;
    }

    return (
        <div className='container login-container'>
            <div className='row'>
            <div className="col-md-12 login-form-2">
                    <h3>Registro</h3>
                    <form onSubmit={ handleRegister }>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Nombre"
                                name='name'
                                value={ name }
                                onChange={ handleInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Correo"
                                name='email'
                                value={ email}
                                onChange={ handleInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contrase単a" 
                                name='password'
                                value={ password }
                                onChange={ handleInputChange }
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repita la contrase単a" 
                                name='password2'
                                value={ password2 }
                                onChange={ handleInputChange }
                            />
                        </div>

                        <div className="form-group">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Crear cuenta" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}
