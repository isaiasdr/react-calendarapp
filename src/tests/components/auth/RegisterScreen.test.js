import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom'; 

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { startRegister } from '../../../actions/auth';
import Swal from 'sweetalert2';

jest.mock('sweetalert2', () => ({
    fire: jest.fn()
}));

jest.mock('../../../actions/auth', () => ({
    startRegister: jest.fn()
})); 

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {};

const store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store } >
        <RegisterScreen />
    </Provider>
);

describe('Tests in <RegisterScreen />', () => {

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should show correctly', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should call the startRegister action', () => {
        wrapper.find('input[name="name"]').simulate('change', { 
            target: { 
                value: 'Pepito', 
                name: 'name' 
            } 
        });

        wrapper.find('input[name="email"]').simulate('change', { 
            target: { 
                value: 'testing1@test.com', 
                name: 'email' 
            } 
        });

        wrapper.find('input[name="password"]').simulate('change', { 
            target: { 
                value: '123456', 
                name: 'password' 
            }
        });

        wrapper.find('input[name="password2"]').simulate('change', { 
            target: { 
                value: '123456', 
                name: 'password2'
            } 
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect( store.dispatch ).toHaveBeenCalled();
        expect( startRegister ).toHaveBeenCalledWith('Pepito', 'testing1@test.com', '123456');
    });

    test('should show the error message', () => {
        wrapper.find('input[name="name"]').simulate('change', { 
            target: { 
                value: 'Pepito', 
                name: 'name' 
            } 
        });

        wrapper.find('input[name="email"]').simulate('change', { 
            target: { 
                value: 'testing1@test.com', 
                name: 'email' 
            } 
        });

        wrapper.find('input[name="password"]').simulate('change', { 
            target: { 
                value: '123456', 
                name: 'password' 
            }
        });

        wrapper.find('input[name="password2"]').simulate('change', { 
            target: { 
                value: '1234567', 
                name: 'password2'
            } 
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        const actions = store.getActions();
        
        expect( store.dispatch ).not.toHaveBeenCalled();
        expect( startRegister ).not.toHaveBeenCalled();
        expect( Swal.fire ).toHaveBeenCalledWith({
            title: 'Error',
            text: 'Las contrase√±as no coinciden',
            icon: 'error'
        });
    });
});
