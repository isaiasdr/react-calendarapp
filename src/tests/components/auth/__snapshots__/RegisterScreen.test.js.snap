// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <RegisterScreen /> should show correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-12 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="name"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="email"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="password2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
